File: tibouchi Watch App/StorylineParser.swift
import Foundation

// MARK: - Erreurs personnalisées

enum StorylineParseError: Error, LocalizedError {
    case invalidJSON
    case missingKey(String)
    case invalidType(String)
    case emptySteps
    case emptyChoicesInStep(Int)

    var errorDescription: String? {
        switch self {
        case .invalidJSON:
            return "Le JSON est invalide ou ne peut pas être décodé."
        case .missingKey(let key):
            return "Clé manquante ou valeur nulle pour la clé : \(key)."
        case .invalidType(let key):
            return "Type invalide pour la clé : \(key)."
        case .emptySteps:
            return "Aucune étape trouvée dans la storyline."
        case .emptyChoicesInStep(let index):
            return "L'étape \(index + 1) ne contient aucun choix."
        }
    }
}

// MARK: - Parseur principal

func parseStoryline(from jsonData: Data) throws -> Storyline {
    print("[StorylineParser] Tentative de décodage du JSON:")
    if let jsonString = String(data: jsonData, encoding: .utf8) {
        print(jsonString)
    } else {
        print("[StorylineParser] Impossible de convertir les données en String.")
    }
    let decoder = JSONDecoder()
    
    do {
        let storyline = try decoder.decode(Storyline.self, from: jsonData)
        print("[StorylineParser] Décodage réussi. Storyline: \(storyline.title)")
        
        guard !storyline.title.isEmpty else {
            print("[StorylineParser] Erreur: titre manquant.")
            throw StorylineParseError.missingKey("title")
        }
        
        let allowedActions = ["nourrir", "abreuver", "aventure", "amour"]
        guard allowedActions.contains(storyline.actionType.rawValue) else {
            print("[StorylineParser] Erreur: actionType invalide: \(storyline.actionType.rawValue)")
            throw StorylineParseError.invalidType("actionType")
        }
        
        guard !storyline.steps.isEmpty else {
            print("[StorylineParser] Erreur: aucune étape trouvée.")
            throw StorylineParseError.emptySteps
        }
        
        for (index, step) in storyline.steps.enumerated() {
            guard !step.prompt.isEmpty else {
                print("[StorylineParser] Erreur: prompt manquant à l'étape \(index + 1)")
                throw StorylineParseError.missingKey("prompt in step \(index + 1)")
            }
            guard !step.choices.isEmpty else {
                print("[StorylineParser] Erreur: aucun choix à l'étape \(index + 1)")
                throw StorylineParseError.emptyChoicesInStep(index)
            }
            for choice in step.choices {
                guard !choice.text.isEmpty else {
                    print("[StorylineParser] Erreur: texte du choix manquant à l'étape \(index + 1)")
                    throw StorylineParseError.missingKey("text in choice (step \(index + 1))")
                }
                guard !choice.consequence.isEmpty else {
                    print("[StorylineParser] Erreur: conséquence du choix manquante à l'étape \(index + 1)")
                    throw StorylineParseError.missingKey("consequence in choice (step \(index + 1))")
                }
                if (choice.gaugeImpact.faim < -10 || choice.gaugeImpact.faim > 10) ||
                   (choice.gaugeImpact.soif < -10 || choice.gaugeImpact.soif > 10) ||
                   (choice.gaugeImpact.mental < -10 || choice.gaugeImpact.mental > 10) {
                    print("[StorylineParser] Erreur: valeur de gaugeImpact hors limites à l'étape \(index + 1)")
                    throw StorylineParseError.invalidType("gaugeImpact values in choice (step \(index + 1))")
                }
            }
        }
        
        return storyline
    } catch let decodingError {
        print("[StorylineParser] Décodage JSON échoué : \(decodingError.localizedDescription)")
        throw StorylineParseError.invalidJSON
    }
}

File: StoryLineGenerator/story_line_parser.py
import json
from typing import Dict, Any
from dataclasses import dataclass
from models import Storyline, ActionType, GaugeState
import uuid

# MARK: - Erreurs personnalisées

@dataclass
class StorylineParseError(Exception):
    """Erreur de base pour le parsing des storylines"""
    message: str

    @classmethod
    def invalid_json(cls) -> 'StorylineParseError':
        return cls("Le JSON est invalide ou ne peut pas être décodé.")
    
    @classmethod
    def missing_key(cls, key: str) -> 'StorylineParseError':
        return cls(f"Clé manquante ou valeur nulle pour la clé : {key}.")
    
    @classmethod
    def invalid_type(cls, key: str) -> 'StorylineParseError':
        return cls(f"Type invalide pour la clé : {key}.")
    
    @classmethod
    def empty_steps(cls) -> 'StorylineParseError':
        return cls("Aucune étape trouvée dans la storyline.")
    
    @classmethod
    def empty_choices_in_step(cls, index: int) -> 'StorylineParseError':
        return cls(f"L'étape {index + 1} ne contient aucun choix.")

# MARK: - Parseur principal

def parse_storyline(json_string: str):
    print("[StorylineParser] Tentative de décodage du JSON:")
    print(json_string)
    
    try:
        data = json.loads(json_string)
        
        print(f"[StorylineParser] ID d'origine dans le JSON: {data.get('id')}")
        # Vérification et génération de l'UUID
        if 'id' not in data:
            print("[StorylineParser] Aucun ID trouvé, génération d'un nouvel UUID")
            data['id'] = str(uuid.uuid4())
        else:
            try:
                uuid_obj = uuid.UUID(data['id'])
                data['id'] = str(uuid_obj)
                print(f"[StorylineParser] UUID valide trouvé et normalisé: {data['id']}")
            except ValueError:
                print(f"[StorylineParser] ID invalide trouvé: {data['id']}, génération d'un nouvel UUID")
                data['id'] = str(uuid.uuid4())
        print(f"[StorylineParser] ID final utilisé: {data['id']}")
        
        corrected_json = json.dumps(data, ensure_ascii=False, indent=2)
        return Storyline(
            id=data['id'],
            title=data['title'],
            action_type=data['action_type'],
            steps=data['steps'],
            raw_json=corrected_json
        ), corrected_json
    except Exception as e:
        raise StorylineParseError(f"Erreur lors du parsing de la storyline : {str(e)}")
